vamb/__main__.py

```diff
@@ -1559,6 +1565,244 @@ def predict_taxonomy(
+        test_mask = ~train_mask
+
+        train_comp = tnfs[train_mask]
+        train_abund = abundance[train_mask]
+        train_lengths = lengths[train_mask]
+        train_targets = targets[train_idx]
+
+        test_comp = tnfs[test_mask]
+        test_abund = abundance[test_mask]
+        test_lengths = lengths[test_mask]
+        test_targets = targets[test_idx]
+
+        model = vamb.taxvamb_encode.VAMB2Label(
+            train_abund.shape[1],
+            len(nodes),
+            nodes,
+            table_parent,
+            nhiddens=[512, 512, 512, 512],
+            hier_loss=taxonomy_options.ploss,
+            cuda=cuda,
+        )
+        dataloader_joint = vamb.taxvamb_encode.make_dataloader_concat_hloss(
+            train_abund,
+            train_comp,
+            train_lengths,
+            train_targets,
+            len(nodes),
+            table_parent,
+            batchsize=taxonomy_options.basic.starting_batch_size,
+        )
+        model.trainmodel(
+            dataloader_joint,
+            nepochs=taxonomy_options.basic.num_epochs,
+            modelfile=None,
+            batchsteps=taxonomy_options.basic.batch_steps,
+        )
+
+        dataloader_joint_test = vamb.taxvamb_encode.make_dataloader_concat_hloss(
+            test_abund,
+            test_comp,
+            test_lengths,
+            test_targets,
+            len(nodes),
+            table_parent,
+            batchsize=taxonomy_options.basic.starting_batch_size,
+        )
+
+        nodes_ar = np.array(nodes)
+        fold_predicted_taxonomies = []
+
+        logger.info(f"Predicting on fold {fold + 1}/{k} with {len(test_idx)} contigs")
+
+        losses = []
+
+        for predicted_vector, predicted_labels, loss in model.predict_with_ground_truth(
+            dataloader_joint_test
+        ):
+            for j in range(predicted_vector.shape[0]):
+                label = predicted_labels[j]
+                while table_parent[label] != -1:
+                    label = table_parent[label]
+                threshold_mask = (
+                    predicted_vector[j] > taxonomy_options.softmax_threshold
+                )
+                pred_ranks = nodes_ar[threshold_mask]
+                ranks = list(pred_ranks[1:])
+                probs = predicted_vector[j][threshold_mask][1:]
+                tax_pred = vamb.taxonomy.PredictedContigTaxonomy(
+                    vamb.taxonomy.ContigTaxonomy(ranks), probs
+                )
+                fold_predicted_taxonomies.append(tax_pred)
+            losses.append(loss.item())
+        return fold_predicted_taxonomies, np.mean(losses)
+
+    fold_args = [
+        (fold, train_idx, test_idx)
+        for fold, (train_idx, test_idx) in enumerate(kf.split(np.arange(n_contigs)))
+    ]
+    loss_tests = []
+
+    with concurrent.futures.ThreadPoolExecutor() as executor:
+        results = list(
+            executor.map(lambda args: train_and_predict_fold(*args), fold_args)
+        )
```

Comment:
I'm not sure this is a good idea. The bottleneck is presumably the neural network training, which is already multithreaded. Running multiple trainings on the same CPU will oversubscribe threads, so it could potentially be less efficient than just running it in serial.
Have you checked that running concurrently is faster? Perhaps the network training can run in serial, and then the benchmarking can run concurrently?

Reply:
I checked, and it was faster. I had the same thoughts, but ended up speeding it up just for my own sake, and kept it like that. Less error prone if serial though, and easier to debug. I think you are right and I move it back to serial for the sake of stability